### PR TITLE
ISPN-9699 0.0.f Global Capacity Factor

### PR DESCRIPTION
**Pull request by karesti**
https://issues.jboss.org/browse/ISPN-9699

depends on https://github.com/infinispan/infinispan/pull/6546

Backport to 9.4.x: https://github.com/infinispan/infinispan/pull/6633

### CODE REVIEW

**Comment by karesti** [comment]
I'm wondering about the security (Secured Actions etc)

File: core/src/main/java/org/infinispan/statetransfer/StateTransferManagerImpl.java [hunk 1]
```diff
@@ -122,6 +125,20 @@ public void rebalance(CacheTopology cacheTopology) {
       }
    }
 
+   private Optional<Float> globalCapacityFactor() {
+      String globalCapacityFactorProperty = System.getProperty(GLOBAL_CAPACITY_FACTOR);
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/statetransfer/StateTransferManagerImpl.java]: As Tristan said, this should be an on-off switch, and it would be better to make it a global setting protected by a feature flag.

File: core/src/main/java/org/infinispan/statetransfer/StateTransferManagerImpl.java [hunk 2]
```diff
@@ -94,14 +94,16 @@ public void start() throws Exception {
          persistentStateChecksum = Optional.empty();
       }
 
+      float capacityFactor = globalConfiguration.isZeroCapacityNode()? 0.0f : configuration.clustering().hash().capacityFactor();
```
Comment (tristantarrant) [inline core/src/main/java/org/infinispan/statetransfer/StateTransferManagerImpl.java]: Formatting (missing a space before the ?

File: core/src/main/resources/schema/infinispan-config-10.0.xsd [hunk 3]
```diff
@@ -266,6 +266,11 @@
         <xs:documentation>Indicates the default cache for this cache container</xs:documentation>
       </xs:annotation>
     </xs:attribute>
+    <xs:attribute name="zero-capacity-node" type="xs:boolean" default="false">
+      <xs:annotation>
+        <xs:documentation>If 'false' then no data is stored in this node. Defaults to 'false'.</xs:documentation>
```
Comment (tristantarrant) [inline core/src/main/resources/schema/infinispan-config-10.0.xsd]: Maybe add a note mentioning that it is equivalent to setting capacity factor 0 on all clustered caches.

File: lock/src/main/java/org/infinispan/lock/impl/ClusteredLockModuleLifecycle.java [hunk 4]
```diff
@@ -66,7 +66,10 @@ public void cacheManagerStarted(GlobalComponentRegistry gcr) {
 
    private static Configuration createClusteredLockCacheConfiguration() {
       ConfigurationBuilder builder = new ConfigurationBuilder();
-      builder.clustering().cacheMode(CacheMode.REPL_SYNC)
+      // ISPN-9699: We want the highest consistency, but we can't apply capacity factor 0 to replicated caches
+      // This is a requirement for the Vert.x Cluster Manager, were we want to have all the caches - public and internal
+      // With capacity factor 0, so vert.x nodes are like light members not containing any data
+      builder.clustering().cacheMode(CacheMode.DIST_SYNC).hash().numOwners(Integer.MAX_VALUE)
```
Comment (tristantarrant) [inline lock/src/main/java/org/infinispan/lock/impl/ClusteredLockModuleLifecycle.java]: Maybe we should introduce the ability to configure CF=0 for members of repl caches which would semantically behave like this but still report as REPL

Comment (karesti) [reply lock/src/main/java/org/infinispan/lock/impl/ClusteredLockModuleLifecycle.java]: This was discussed on the meeting, and we decided to use a DIST cache on locks with max num of owners because for replicated caches apparently is not that simple ... @danberindei @wburns @pruivo any thoughts ?

File: core/src/main/resources/schema/infinispan-config-10.0.xsd (hunk 3 above)
Comment (karesti) [reply core/src/main/resources/schema/infinispan-config-10.0.xsd]: It's not exactly true, because internal caches won't have capacity factor 0, and with this modification all caches are capacity factor 0

Comment (pruivo) [reply core/src/main/resources/schema/infinispan-config-10.0.xsd]: do you mean `If true`?

File: core/src/test/java/org/infinispan/distribution/ZeroCapacityNodeTest.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+import static org.testng.AssertJUnit.assertEquals;
+
+/**
+ * Test the capacity factor for lite instance
+ *
+ * @author Katia Aresti
+ * @since 9.4
+ */
+@Test(groups = "functional", testName = "distribution.ch.ZeroCapacityNodeTest")
+public class ZeroCapacityNodeTest extends MultipleCacheManagersTest {
```
Comment (pruivo) [inline core/src/test/java/org/infinispan/distribution/ZeroCapacityNodeTest.java]: why extending `MultipleCacheManagersTest`?

File: core/src/test/java/org/infinispan/distribution/ZeroCapacityNodeTest.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+   public static final int NUM_SEGMENTS = 60;
+
+   @Override
+   protected void createCacheManagers() throws Throwable {
+      // Do nothing here, create the cache managers in the test
+   }
+
+   public void testCapacityFactorForLiteInstance() {
```
Comment (pruivo) [inline core/src/test/java/org/infinispan/distribution/ZeroCapacityNodeTest.java]: I was expecting more tests to be honest. 
* Does both cache mode behaves correctly with zero and non-zero nodes? 
* Will counter and multimap work correctly? 
* what happens if you start a cache in a non-zero node? Will all the zero nodes start the cache as well?

File: core/src/test/java/org/infinispan/distribution/ZeroCapacityNodeTest.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+      cb.clustering().cacheMode(CacheMode.DIST_SYNC);
+      cb.clustering().hash().numSegments(NUM_SEGMENTS);
+
+      cb.clustering().hash().capacityFactor(0.5f);
+      addClusterEnabledCacheManager(gc, cb);
+      waitForClusterToForm();
+      assertCapacityFactors(0.0f);
+   }
+
+   private void assertCapacityFactors(float... expectedCapacityFactors) {
```
Comment (pruivo) [inline core/src/test/java/org/infinispan/distribution/ZeroCapacityNodeTest.java]: why an array?

File: lock/src/main/java/org/infinispan/lock/impl/ClusteredLockModuleLifecycle.java (hunk 4 above)
Comment (pruivo) [reply lock/src/main/java/org/infinispan/lock/impl/ClusteredLockModuleLifecycle.java]: @tristantarrant the replicated mode uses a simplified version of `ConsistentHashs` and totally ignores the CF and `num-owners`. I guess it can be done but we had to drop the `ReplicatedConsistentHash` and do some tunning probably (or replace it with the `DefaultConsistentHash` with `num-owners == MAX_INTEGER`.

File: core/src/main/java/org/infinispan/configuration/global/GlobalConfigurationBuilder.java [hunk 8]
```diff
@@ -157,6 +161,11 @@ public ShutdownConfigurationBuilder shutdown() {
       return modules;
    }
 
+   public GlobalConfigurationBuilder zeroCapacityNode(Boolean zeroCapacityNode) {
```
Comment (wburns) [inline core/src/main/java/org/infinispan/configuration/global/GlobalConfigurationBuilder.java]: Should be boolean, otherwise NPE will be thrown.

File: core/src/main/java/org/infinispan/configuration/parsing/ConfigurationBuilderHolder.java [hunk 9]
```diff
@@ -50,11 +50,7 @@ public ConfigurationBuilder getCurrentConfigurationBuilder() {
    }
 
    public ConfigurationBuilder getDefaultConfigurationBuilder() {
-      if (globalConfigurationBuilder.defaultCacheName().isPresent()) {
-         return namedConfigurationBuilders.get(globalConfigurationBuilder.defaultCacheName().get());
-      } else {
-         return null;
-      }
+       return globalConfigurationBuilder.defaultCacheName().map(namedConfigurationBuilders::get).orElse(null);
```
Comment (wburns) [inline core/src/main/java/org/infinispan/configuration/parsing/ConfigurationBuilderHolder.java]: Technically this is another lambda and out of scope of this PR imo.

File: core/src/main/java/org/infinispan/configuration/parsing/Parser.java [hunk 10]
```diff
@@ -553,6 +553,10 @@ private void parseContainer(XMLExtendedStreamReader reader, ConfigurationBuilder
                builder.shutdown().hookBehavior(ShutdownHookBehavior.valueOf(value));
                break;
             }
+            case ZERO_CAPACITY_NODE: {
+               builder.zeroCapacityNode(Boolean.valueOf(value));
```
Comment (wburns) [inline core/src/main/java/org/infinispan/configuration/parsing/Parser.java]:
```suggestion
               builder.zeroCapacityNode(Boolean.parseBoolean(value));
```

File: core/src/test/java/org/infinispan/distribution/ZeroCapacityNodeTest.java (hunk 7 above)
Comment (karesti) [reply core/src/test/java/org/infinispan/distribution/ZeroCapacityNodeTest.java]: refactored, the other was a test I took inspiration from but not working one

File: core/src/main/java/org/infinispan/statetransfer/StateTransferManagerImpl.java [hunk 11]
```diff
@@ -94,14 +94,16 @@ public void start() throws Exception {
          persistentStateChecksum = Optional.empty();
       }
 
+      float capacityFactor = globalConfiguration.isZeroCapacityNode() ? 0.0f : configuration.clustering().hash().capacityFactor();
```
Comment (wburns) [inline core/src/main/java/org/infinispan/statetransfer/StateTransferManagerImpl.java]: I really wish this was set in the `HashConfigurationBuilder#create` method instead, so we don't have to worry about all the references. But we don't have access to the `GlobalConfiguration` there as far as I know. :(

File: core/src/main/java/org/infinispan/configuration/parsing/ConfigurationBuilderHolder.java (hunk 9 above)
Comment (karesti) [reply core/src/main/java/org/infinispan/configuration/parsing/ConfigurationBuilderHolder.java]: ah yes, haha I saw it and I thought in the refactoring but can remove it without any problem.
I forgot I did it here

File: core/src/main/java/org/infinispan/configuration/global/GlobalConfigurationBuilder.java (hunk 8 above)
Comment (karesti) [reply core/src/main/java/org/infinispan/configuration/global/GlobalConfigurationBuilder.java]: yes

File: core/src/main/java/org/infinispan/configuration/parsing/Parser.java (hunk 10 above)
Comment (karesti) [reply core/src/main/java/org/infinispan/configuration/parsing/Parser.java]: I was using valueOf as for the statistics, but parseBoolean is the same

**Comment by karesti** [comment]
There is a test failure that does not seem related. @pruivo do you think that we really need additional tests for counters and locks?

**Comment by karesti** [comment]
@pruivo I'm going to add a test on locks

**Comment by karesti** [comment]
@pruivo added some tests on counters and detected a problem with locks. The zero capacity node is counted in the cluster as a non owner, so if we say we need Max Integer values owners, cluster won't start. I think this must be set up depending on the configuration. By default I say 2, and I have this issue ISPN-9625 to make locks xml configurable (such as how should they behave in when there is a split brain)

File: core/src/main/java/org/infinispan/configuration/global/GlobalConfigurationBuilder.java [hunk 12]
```diff
@@ -15,6 +15,8 @@
 
 public class GlobalConfigurationBuilder implements GlobalConfigurationChildBuilder {
 
+   public static final String ZERO_CAPACITY_NODE_FEATURE = "zero-capacity-node";
```
Comment (wburns) [inline core/src/main/java/org/infinispan/configuration/global/GlobalConfigurationBuilder.java]: I am not liking this here. Especially if this feature is removed earlier. I would prefer if this variable was defined on a class that isn't marked as public.

File: core/src/main/java/org/infinispan/configuration/global/GlobalConfigurationBuilder.java [hunk 13]
```diff
@@ -157,6 +161,11 @@ public ShutdownConfigurationBuilder shutdown() {
       return modules;
    }
 
+   public GlobalConfigurationBuilder zeroCapacityNode(boolean zeroCapacityNode) {
```
Comment (wburns) [inline core/src/main/java/org/infinispan/configuration/global/GlobalConfigurationBuilder.java]: Seems to be missing some documentation. Also I don't remember the discussions, but is this going to be fully supported mode? I wonder if we should add an @Experimental tag to it?

Comment (karesti) [reply core/src/main/java/org/infinispan/configuration/global/GlobalConfigurationBuilder.java]: what do you mean by fully supported?

File: core/src/main/java/org/infinispan/configuration/global/GlobalConfigurationBuilder.java (hunk 12 above)
Comment (karesti) [reply core/src/main/java/org/infinispan/configuration/global/GlobalConfigurationBuilder.java]: I moved the constant to be private and if I need to access from somewhere else I will find another place

File: core/src/test/java/org/infinispan/distribution/ZeroCapacityNodeTest.java (hunk 5 above)
Comment (karesti) [reply core/src/test/java/org/infinispan/distribution/ZeroCapacityNodeTest.java]: @pruivo I need to test that in a cluster I have a node with zero capacity and it works

File: core/src/main/java/org/infinispan/statetransfer/StateTransferManagerImpl.java (hunk 11 above)
Comment (karesti) [reply core/src/main/java/org/infinispan/statetransfer/StateTransferManagerImpl.java]: Can I do something here now then ?

File: core/src/main/java/org/infinispan/configuration/global/GlobalConfigurationBuilder.java (hunk 13 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/configuration/global/GlobalConfigurationBuilder.java]: I am saying as in it doesn't have @Experimental tag on it and is shown to users and properly documented everything. For example this method has no javadoc on it, doesn't tell me what this means at all.

Comment (karesti) [reply core/src/main/java/org/infinispan/configuration/global/GlobalConfigurationBuilder.java]: @wburns I added some extra javadoc in a commit

Comment (karesti) [reply core/src/main/java/org/infinispan/configuration/global/GlobalConfigurationBuilder.java]: @wburns some java doc added in another commit

**Comment by wburns** [comment]
Integrated into master, thanks @karesti !
